File: Day-03/python/RainbowDasLabs/day_3.py
values = [e.strip() for e in open("puzzle.txt") if e.strip()]


def common_bit(val: list[str], index: int, method: str = "most") -> int:
    zero_common = sum([int(e[index]) for e in val]) < len(val) / 2
    return (0 if zero_common else 1) if method == "most" else (1 if zero_common else 0)


def calc_bits(val: list[str], method: str = "most") -> str:
    return "".join([str(common_bit(val, i, method)) for i in range(len(val[0]))])


print(f"Part one {int(calc_bits(values, 'most'), 2) * int(calc_bits(values, 'least'), 2)}")


def bit_filter(val: list[str], index: int, method: str):
    if len(val) == 1:
        return val[0]
    bit = common_bit(val, index, method)
    return bit_filter([e for e in val if int(e[index]) == bit], index + 1, method)


print(f"Part two {int(bit_filter(values, 0, 'most'), 2) * int(bit_filter(values, 0, 'least'), 2)}")

File: Day-03/java/RainbowDashLabs/Day3.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Day3 {
    private static int toNum(String str, int index){
        return Character.getNumericValue(str.charAt(index));
    }

    private static int commonBit(List<String> val, int index, Method method) {
        var zeroCommon = val.stream().mapToInt(e -> toNum(e, index)).sum() < val.size() / 2.0;
        return method == Method.MOST ? zeroCommon ? 0 : 1 : zeroCommon ? 1 : 0;
    }

    private static String calcBits(List<String> val, Method method) {
        return IntStream.range(0, val.get(0).length())
                .mapToObj(i -> String.valueOf(commonBit(val, i, method)))
                .collect(Collectors.joining());
    }

    private static String bitFilter(List<String> val, int index, Method method) {
        if (val.size() <= 1) return val.get(0);
        var bit = commonBit(val, index, method);
        return bitFilter(val.stream().filter(e -> toNum(e, index) == bit).collect(Collectors.toList()), index + 1, method);
    }

    public static void main(String[] args) throws IOException {
        var input = Files.readAllLines(Path.of("puzzle.txt"));

        System.out.printf("Part one %s%n", Integer.parseInt(calcBits(input, Method.MOST), 2) * Integer.parseInt(calcBits(input, Method.LEAST), 2));

        System.out.printf("Part two %s%n", Integer.parseInt(bitFilter(input, 0, Method.MOST), 2) * Integer.parseInt(bitFilter(input, 0, Method.LEAST), 2));
    }

    private enum Method {
        LEAST, MOST
    }
}
